### ISSUE TITLE
Include "mandatory QueryParam" (QueryParams, QueryFlag as well) in the next release

### ISSUE THREAD

**Issue by alpmestan** [haskell-servant/servant#241]
Over the past few weeks, several users mentionned that they would like a variation of `QueryParam`s that makes them mandatory. That is, instead of handing a `Maybe a` to the handlers (hence giving Nothing when the query string param is absent or can't be decoded), this variation would hand an `a` directly and just error out if the parameter is absent or can't be decoded.

@kosmikus also mentionned it would be nice to differentiate between "not there" and "can't be decoded", even though this one is less of a problem because you can always just get a `Text` to your handler and let it decide what to do when decoding fails. Whereas today the only way to get a "mandatory `QueryParam`" is to write a new combinator for it.

I see two ways to do this:
- Path of least code breakage: introduce new combinators, "mandatory" variants of the existing ones. Regarding the naming convention, we could add a `'` to the name or some other thing.
- Change the combinators to something like:

``` haskell
data ParamType = Optional | Mandatory
data QueryParam (name :: Symbol) a (paramtype :: ParamType)
```

and have the `HasServer`/`Server` instances dispatch on `paramtype` to get the current behavior or the "mandatory" one.

**Comment by jkarni** [comment]
> Whereas today the only way to get a "mandatory QueryParam" is to write a new combinator for it.

Well, you can also just 400/404 on `Nothing` in your handler.

**Comment by edsko** [comment]
> @kosmikus also mentionned it would be nice to differentiate between "not there" and "can't be decoded", even though this one is less of a problem because you can always just get a Text to your handler and let it decide what to do when decoding fails. 

I disagree with this rather strongly. The whole point is to specify the API once and for all in the types; I don't want to have to check for presence of arguments, nor check for their well-formedness. This is why servant is so nice -- we give a specification, and don't worry about such annoying details anywhere else. A parse error in a query argument should result in a 500 (or whatever), without anything to do on my part (just like for request bodies).

**Comment by jkarni** [comment]
> The whole point is to specify the API once and for all in the types

I agree. On the other hand, it's not clear to me how just passing a sum type with three constructors instead of three (in the `Optional` case) won't also be encouraging people to throw errors from their handlers. What could they want to do with a malformed query param besides throw an error, or treat it as a missing one?

**Comment by edsko** [comment]
I'm not sure if the above was in answer to me, but as far as I'm concerned, the handler should _not_ have to deal with malformed arguments, nor deal with missing arguments. Both should automatically be handled, and probably throw a 500 Bad Request or something similar. 

For the front-end designer however it is incredibly useful if the difference between missing and illformed is clear. Distinguishing "Argument such and such missing" or "Could not parse argument such and such" is very very useful for the front end designer.

**Comment by edsko** [comment]
Perhaps this is what thte confusion is stemming from. The _handler_ does not need to be able to distinguish between missing and illformed; in both cases, the handler wouldn't even be invoked. But we just have to make sure that error that gets generated (by servant) _does_ distinguish between them.

**Comment by jkarni** [comment]
Ok, that makes sense. Maybe we don't even need the `Required` flag, then? We could just let people decode as `Maybe a` (or `Either`) if they want it to be optional.

**Comment by edsko** [comment]
That would make a lot of sense to me, provided that this fits with the decoding scheme (such that `Nothing` can be represented by the absence of the parameter).

**Comment by edsko** [comment]
Perhaps introduce a `Maybe`-isomorphic type for that specific purpose, in case somebody wants to have an explicit representation of `Nothing` versus `Just` by an argument? (Not sure why anybody would want that, but it's not inconceivable.)

**Comment by jkarni** [comment]
> in case somebody wants to have an explicit representation of Nothing versus Just by an argument

Sorry, I didn't quite understand this.

**Comment by edsko** [comment]
Well, normally it's the users (or servant) decision how parameters are encoded, no? Perhaps they prefer 'Nothing' to be encoded by a specific value of a (present) parameter, rather than by an absent parameter. If you use Maybe to mean optional, you are hardcoding the behaviour of Maybe. It may or may not be an issue. (To me personally it wouldn't matter.)

**Comment by jkarni** [comment]
Ah, I see. Very fair. We could use that type in other decoding scenarios too.

**Comment by alpmestan** [comment]
Hmm, I think this might be a problem for me. I usually reach for `QueryParam` for parameters that are *optional* (i.e I usually have a default value for them if one is not provided, or I can do without any value). I don't want *servant* to systematically error out whenever the specified value is illformed or when the param is absent! You can technically reach the endpoint without specifying query params, by just making a request to the path and stripping away the `?` and anything that comes after in the URL. So I'm not happy with the proposed behavior. I'm hoping we can meet in the middle and support both scenarios.

**Comment by edsko** [comment]
You can technically reach an endpoint irrespective of the request body too -- but we don't put the handler in charge of decoding it or checking if its there. Of course, truly _optional_ parameter must be supported; I think the proposal is to model them by something like `QueryParam (Optional a)`, and indeed you could model an optional request body the same way `ReqBody (Optional a)`. Present-but-malformed parameters are a different story though; if you really want the handler to be reachable in that case I guess the only option you would have with this proposal is to model them as `ReqBody (Optional Text)`, and deal with the decoding yourself.

**Comment by alpmestan** [comment]
So the idea is to have `Server (QueryParam sym a :> blah) = a -> Server blah`, with:

``` haskell
data Optional a = NotThere | There a
-- along with Functor, Applicative, Monad, Foldable, Traversable instances
```

that would let us add optional semantics to a query param? Doesn't it force us to have overlapping instances for every combinator we might want to use `Optional` with though? While the `Server` definitions might be uniform, I'm not sure we can share common definitions for `route` (`HasServer`'s method), can we?

**Comment by kosmikus** [comment]
Here's my interpretation of the discussion. Currently, we have
```haskell
class FromText a where
  fromText :: Text -> Maybe a
```
and
```haskell
instance (KnownSymbol sym, FromText a, HasServer sublayout)
      => HasServer (QueryParam sym a :> sublayout) where
```
The function `fromText` is only called when the parameter is both present and has a value. If `fromText` fails, it is treated equivalently to the parameter being missing or not having a value. The handler has to decide what to do with any kind of "failure" or the result.

I think we should keep the `HasServer` instance, but change `FromText` to
```haskell
class FromText a where
  fromText :: Param -> Decoded a
```
(which may require changing the name of `FromText`) where
```haskell
data Param = Missing | NoValue | WithValue Text
data Decoded a = Required | Illegal | Accepted a
```
We'd now *always* call `fromText` on the parameter. The `fromText` function can decide whether it wants to require a value or not, and can return a meaningful result such as `Required` or `Illegal`, possibly with additional error messages attached. These would automatically lead to error codes being returned. The handler would only be called for an `Accepted` value, with an `a`.

If we'd want to use this to implement an *optional* query parameter, we'd define an instance such as
```haskell
instance FromText (Maybe MyData) where
  fromText Missing = Accepted Nothing
  fromText NoValue = Accepted Nothing
  fromText (WithValue txt) = case decodeMyData of
    Nothing -> Illegal
    Just x -> Accepted (Just x)
```

If we'd want to use this to implement a *required* query parameter, we'd say
```haskell
instance FromText MyData where
  fromText Missing = Required
  fromText NoValue = Illegal
  fromText (WithValue txt) = case decodeMyData txt of
    Nothing -> Illegal
    Just x -> Accepted x
```

**Comment by alpmestan** [comment]
`FromText` is also used for `Capture`s. The proposed change for this typeclass would mess things up for `Capture`. It would then have to get its own typeclass...?

**Comment by kosmikus** [comment]
That's why I said we might want to change the name. This one could coexist with `FromText` used for captures, but it would not be the same.

**Comment by jkarni** [comment]
The situation is changed somewhat by #250 .

**Comment by fizruk** [comment]
I have managed to create a `Required` combinator, which allows to
- keep all the existing combinators;
- "require" any `a`, if `ServerT (a :> sub) = Maybe x -> ServerT sub` (for some x);
- this makes `Required` applicable to any user-defined combinators as well.

Usage is simple and looks like this:

```haskell
type TestApi
    = "optional" :> QueryParam "param1" Int            :> Get '[JSON] (Maybe Int)
 :<|> "required" :> Required (QueryParam "param2" Int) :> Get '[JSON] Int

server :: Server TestApi
server = handler1 :<|> handler2
  where
    handler1 = return
    handler2 = return
```

Unfortunately, `HasServer`'s definition currently makes life hard, so implementation required a few "hacks":
- closed type families to get `a` from `Maybe a -> b`;
- `throw`/`catch` exception to detect missing parameter.

These hacks are local though, so it might be okay.

```haskell
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE UndecidableInstances #-} -- this is due to nested type families
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE DeriveDataTypeable #-}
module Test where

import Control.Exception
import Data.Typeable

import Servant
import Servant.Server.Internal

-- | Makes any optional server parameter required.
data Required a

-- | Get the @a@ from @Maybe a -> b@.
type family ParamType t :: * where
  ParamType (Maybe a -> b) = a

-- | This exception is used locally to detect missing parameter.
data MissingParameter = MissingParameter deriving (Show, Typeable)
instance Exception MissingParameter

-- | The second constraint makes sure that
--
-- @
-- Server (a :> sub) = Maybe x -> Server sub
-- @
--
-- for some @x@.
instance
  ( HasServer (a :> sub)
  , Server (a :> sub) ~ (Maybe (ParamType (Server (a :> sub))) -> Server sub))
    => HasServer (Required a :> sub) where

  type ServerT (Required a :> sub) m
    = ParamType (ServerT (a :> sub) m) -> ServerT sub m

  route _ subserver request respond =
    route (Proxy :: Proxy (a :> sub)) (withRequired subserver) request respond
    `catch` (\(e :: MissingParameter) -> respond $ failWith NotFound)
    where
      -- we don't know the type of f here,
      -- so we throw exception in case of missing parameter
      withRequired f Nothing  = throw MissingParameter
      withRequired f (Just x) = f x
```

**Comment by fizruk** [comment]
Any thoughts? Does this solve the problem discussed here or did I miss something?

**Comment by jkarni** [comment]
I think it'd be nicer to start using type synonyms. Something like:
```
data Query required typ
type QueryFlag = Query NotRequired Bool
type QueryParam a = Query NotRequired a

data Required = Required | NotRequired

data Present a = Present a | Absent

type family QueryArg r a where
   QueryArg Required a = a
   QueryArg NotRequired a = Present a

instance (HasServer sub) => HasServer (Query (r :: Required) a :> sub where
   type ServerT (Query r a :> sub) m = QueryArg r a :> ServerT sub m
```

When the argument is `Required` and the param is not present, we throw a 400. If the decoding fails, we also throw a 400. If the type (second arg of `Query`) is `Maybe a`), decoding always succeeds, so the handler can decide what to do with malformed params. This, I think, makes it easy (indeed, the default) for parameters to be optional (in the sense that the handler still gets called if they are entirely absent), but it also makes it the default for parameters that are malformed to constitute an error. I think this would also allow us to avoid throwing and catching exceptions.

**Comment by fizruk** [comment]
What I don't like about current type synonyms approach is that it does not easily extend to other combinators (e.g. `Header`) or user-defined ones.

> I think this would also allow us to avoid throwing and catching exceptions.

Is there anything else pushing you from `Required` combinator I proposed?

**Comment by jkarni** [comment]
I think your solution is neat insofar as it's pretty general, and it's a very positive point in its favour, testifying to its modularity, that it doesn't actually involve code change in the rest of the codebase. But part of my concern is precisely that generality. This would for example work with a `ReqBody ... (Maybe a)`, with captures, with query params, with headers... But the errors shouldn't all be 400 or all 404. Indeed, the response bodies themselves perhaps should be different. 

Moreover, I think there should be a clearer distinction between missing and not decodable. I don't think this distinction always makes sense, though (for captures definitely not, and I think for request bodies not either). So again I don't know if this solution will generalize well to all cases.

There is also the more general design question underlying this. When there are a bunch of related combinators with slightly different behaviour, do we want type synonyms? I think we have been discussing that and are thinking of moving in that direction. Both type synonyms and your solution solve this particular problem. Ideally, we would have the fewest possible different mechanisms in operation, so the codebase is easier to understand and the library easier to use. Admittedly, so far we haven't been using type synonyms very much, so right now its a toss.

I'm completely open to further discussion though :)

**Comment by donatello** [comment]
As a user of this library, I would very much like to use a combinator that differentiates between the absence of a query parameter and a failure to covert it to the desired type. Right now, I am resorting to parsing the query parameter to `Text` and then converting it to my desired type (and returning a failure response if the conversion fails).

I would very much appreciate it, if you could provide a demonstration of writing a combinator that does this in the currently released version (or maybe you could point me to an implementation somewhere).

**Comment by gasi** [comment]
> As a user of this library, I would very much like to use a combinator that differentiates between the absence of a query parameter and a failure to covert it to the desired type.

I feel the same and would love to see a proposal on this front. I have a few use cases like that and would be more than happy to help test the design/implementation.

**Comment by gmalecha** [comment]
It seems like this situation is not resolved and looks fairly inactive. Are there any hacks to get around this problem? For my purposes, I only need to query an API, so I looked into making my own instance of `HasLink` but `Link` is hidden preventing me from making a new instance for `HasLink (ReqQueryParam ... :> sub)`.

**Comment by phadej** [comment]
We can export `Link` from internal module. PR welcome

**Comment by stiell** [comment]
I found it easiest to implement a mandatory variant of `QueryParam` by using `QueryParam` itself in the `HasLink` and `HasClient` instances. This requires no access to servant internals:

```hs
type role QueryParam' phantom phantom
data QueryParam' (sym :: Symbol) (a :: k)

instance (KnownSymbol sym, ToHttpApiData v, HasLink sub) =>
         HasLink (QueryParam' sym v :> sub) where
  type MkLink (QueryParam' sym v :> sub) = v -> MkLink sub
  toLink _ l v = toLink (Proxy :: Proxy (QueryParam sym v :> sub)) l (Just v)

instance (KnownSymbol sym, ToHttpApiData a, HasClient api) =>
         HasClient (QueryParam' sym a :> api) where
  type Client (QueryParam' sym a :> api) = a -> Client api
  clientWithRoute Proxy req param =
    clientWithRoute (Proxy :: Proxy (QueryParam sym a :> api)) req (Just param)
```

**Comment by gmalecha** [comment]
Thanks, @stiell.

**Comment by Ericson2314** [comment]
@stiell's example doesn't include the `HasServer` instance. That one is a bit more involved I think---I couldn't figure how to leverage `QueryParam`'s instance for a desguaring for example.

**Comment by alpmestan** [comment]
Right I'm not sure you can just rely on the existing instance here. @stiell Do you have the `HasServer` instance or do we need to write it? Then someone could include this in servant-contrib or in the core servant packages. Well, ideally with instances for the other interpretations too... =)

**Comment by phadej** [comment]
solved by #873
